upup/pkg/kutil/rollingupdate_cluster_validate_drain.go

```diff
@@ -0,0 +1,313 @@
+	}
+}
+
+// RollingUpdateDrainValidate performs a rolling update on a list of ec2 instances.
+func (n *CloudInstanceGroup) RollingUpdateDrainValidate(rollingUpdateData *RollingUpdateDataDrainValidate) error {
+
+	// we should not get here, but hey I am going to check
+	if rollingUpdateData == nil {
+		return fmt.Errorf("rollingUpdate cannot be nil")
+	}
+
+	// Do not need a k8s client if you are doing cloud only
+	if rollingUpdateData.K8sClient == nil && !rollingUpdateData.CloudOnly {
+		return fmt.Errorf("rollingUpdate is missing a k8s client")
+	}
+
+	if rollingUpdateData.InstanceGroupList == nil {
+		return fmt.Errorf("rollingUpdate is missing a the InstanceGroupList")
+	}
+
+	c := rollingUpdateData.Cloud.(awsup.AWSCloud)
+
+	update := n.NeedUpdate
+	if rollingUpdateData.Force {
+		update = append(update, n.Ready...)
+	}
+
+	if !rollingUpdateData.IsBastion && rollingUpdateData.FailOnValidate && !rollingUpdateData.CloudOnly {
+		_, err := validate.ValidateCluster(rollingUpdateData.ClusterName, rollingUpdateData.InstanceGroupList, rollingUpdateData.K8sClient)
+		if err != nil {
+			return fmt.Errorf("cluster %s does not pass validation", rollingUpdateData.ClusterName)
+		}
+	}
+
+	for _, u := range update {
+
+		if !rollingUpdateData.IsBastion {
+			if rollingUpdateData.CloudOnly {
+				glog.Warningf("Not draining nodes - cloud only is set.")
+			} else {
+
+				drain, err := NewDrainOptions(nil, u.Node.ClusterName)
+
+				if err != nil {
+
+					glog.Warningf("Error creating drain: %v.", err)
+					if rollingUpdateData.FailOnDrainError {
+						return fmt.Errorf("error creating drain: %v.", err)
+					} else {
+						glog.Infof("Proceeding with rolling-update since fail-on-drain-error is set to false.")
+					}
+
+				} else {
+
+					err = drain.DrainTheNode(u.Node.Name)
+					if err != nil {
+						glog.Warningf("Error draining node: %v.", err)
+						if rollingUpdateData.FailOnDrainError {
+							return fmt.Errorf("error draining node: %v", err)
+						} else {
+							glog.Infof("Proceeding with rolling-update since fail-on-drain-error is set to false.")
+						}
+					}
+
+				}
+			}
+		}
+
+		instanceID := aws.StringValue(u.ASGInstance.InstanceId)
+		glog.Infof("Stopping instance %q in AWS ASG %q.", instanceID, n.ASGName)
+
+		request := &ec2.TerminateInstancesInput{
+			InstanceIds: []*string{u.ASGInstance.InstanceId},
+		}
+		_, err := c.EC2().TerminateInstances(request)
+		if err != nil {
+			return fmt.Errorf("error deleting instance %q: %v", instanceID, err)
+		}
+
+		if !rollingUpdateData.IsBastion {
+
+			// Wait for new EC2 instances to be created
+			time.Sleep(rollingUpdateData.Interval)
+
+			if rollingUpdateData.CloudOnly {
+				glog.Warningf("Not validating nodes as cloudonly flag is set.")
+				return nil
+			}
+
+			// Wait until the cluster is happy
+			for i := 0; i <= retries; i++ {
+				_, err = validate.ValidateCluster(rollingUpdateData.ClusterName, rollingUpdateData.InstanceGroupList, rollingUpdateData.K8sClient)
```

Comment:
I think it checks that all nodes are ready.  But it doesn't check that the new node has registered with the API.  If the rolling-update meant that all new instances failed to register with the API server, would we stop the rolling-update?